addons/rmsmartshape/shapes/shape_base.gd

```diff
@@ -339,8 +339,8 @@ func _update_curve(p_array: SS2D_Point_Array) -> void:
 
 func get_vertices() -> PackedVector2Array:
 	var positions: PackedVector2Array = []
-	for p_key in __points.get_all_point_keys():
-		positions.push_back(__points.get_point_position(p_key))
+	for p_key in get_all_point_keys():
+		positions.push_back(get_point_position(p_key))
```

Comment:
The idea behind replacing as many occurrences of directly accessing `__points` as possible, was to a) make it simpler, b) use existing methods, thus avoiding code duplication, and c) avoid accidentally using `_points` by relying completely on those existing helpers.


> On my system (relatively high spec) 1000 function calls consume up to 0.1 ms.

Have you tested that on 3.x or 4.0?


> will hurt performance a bit.

I think it's alright. 0.1ms isn't relevant in this case, especially, since this is not computed 60+ times per second but once when the shape regenerates. Also in terms of performance there are other more severe parts that could need optimizations.

Reply:
> Have you tested that on 3.x or 4.0?

Tested on 4.0.

> I think it's alright. 0.1ms negligible in this case, especially since this is not computed 60+ times per second but once when the shape regenerates. Also in terms of performance there are other more severe parts that could need optimizations.

get_vertices() is actually used in quite a few places and is computed during `process_frame` when you are holding LMB and moving a point in the editor. Mesh is recomputed in real-time. The lags with 300+ points are quite epic btw (3x).

Reply:
Then the bottleneck is not this line but `get_vertices()` not being cached + tesselation not being cached + probably various other parts that are not cached or inefficiently implemented.

Reply:
I also noticed heavy lags using this plugin in 3.x and it could really profit of some optimization, but this is not the part to start.
Manually inlining function calls here to save 0.1ms in some cases is not worth it IMO, especially when there are other, more severe bottlenecks.

Reply:
> The idea behind replacing as many occurrences of directly accessing `__points` as possible, was to a) make it simpler, b) use existing methods, thus avoiding code duplication, and c) avoid accidentally using `_points` by relying completely on those existing helpers.

Anyone can still write code that uses `_points` - it's not really solving the problem. The problem is that we have 3 ways of accessing the same thing and one of them is dangerous - but only when you do it from a setter.

Reply:
Actually - 4 ways (see get_point_array) :rofl:

Reply:
True, but it streamlines how to access that property. So instead of having multiple lines accessing that property all in a different way, we have one function that does the job that is used throughout the code.

Reply:
How about we remove __points altogether. Convert _points into property without setters and getters. Leave the public interface as such and cut at least 2 ways of accessing this mess. Then there is no problem. Initializing part from the property could move into _ready().

Reply:
My point is that property shouldn't be a property in the first place (conventionally speaking, i.e. having setter and  getter).

Reply:
Also, exposing points resource to the editor we allow users to mess with internals in a way that could break stuff easily. It can be serialized without showing it in inspector. And people can still access that data though editing the scene file (only needed if something breaks terribly).

Reply:
But without a setter, the shape would not update when setting a new point array. E.g. when the engine loads the shape and sets the point array, it will not update.

Reply:
> But without a setter, the shape would not update when setting a new point array. E.g. when the engine loads the shape and sets the point array, it will not update.

That logic goes into _ready().

Reply:
But then we have an exported property that can be modified in the inspector, but does not reflect those changes because there is no setter. That's rather unintuitive.

In the end that's the way to go. We need a private backing field and an accessor property. It's ugly but unless Godot supports a way of serializing the points without the user knowing, that's the only way.

Now that I think about it, code should use `_points` everywhere instead of `__points`, except for preventing recursion, because that's the public property that should be used.

Reply:
Looking back at 2183d38, do we actually need the make_unique flag in set_point_array? As I see it, that's the only reason we need the accessor property, to call set_point_array with make_unique = true.
If we don't need it, we could remove it, use set_point_array directly as setter, and we're done.

Reply:
Godot does support serializing the points without user knowing. It's not heavily advertised.
Snippet from 3.x:
```
func _get_property_list():
	# Export 'children' property, but hide it from inspector
	var serializable = [
		{
			name = "children",
			usage = PROPERTY_USAGE_NOEDITOR,
			type = TYPE_ARRAY,
		},
	]
	return serializable
```

Reply:
Hm, that might be another solution then.

Reply:
> Looking back at [2183d38](https://github.com/mphe/SmartShape2D/commit/2183d38cfd29577078c2b659a3c422f4ed820a83), do we actually need the make_unique flag in set_point_array? As I see it, that's the only reason we need the accessor property, to call set_point_array with make_unique = true.
> If we don't need it, we could remove it, use set_point_array directly as setter, and we're done.

Indeed, the whole dance started because of that parameter. The only place that parameter is used is in Shape_Meta it seems.

Reply:
And shape_meta is also a candidate for removal IMO, as it serves no purpose than for duplicating a shape

Reply:
I wonder if anyone uses it. I didn't even know it exists until port.

Reply:
> And shape_meta is also a candidate for removal IMO, as it serves no purpose than for duplicating a shape, which could be implemented differently.

Wait, why are points duplicated (made unique) in a setter in the first place? That will actually generate unique id in 4.0 every time!
This means, that each time anyone opens a scene with a shape a new instance of points will be created, if that scene is saved (or run) it will produce a file with essentially same content but a different id for points resource. As such, it will have a different checksum and generate change in the working tree for GIT.

Reply:
I guess the intent was to forcefully prevent two or more shapes sharing points.
But generating different IDs is bad and must be fixed. However, that would be a new PR.

Reply:
> The idea behind replacing as many occurrences of directly accessing `__points` as possible, was to a) make it simpler, b) use existing methods, thus avoiding code duplication, and c) avoid accidentally using `_points` by relying completely on those existing helpers.

a) we should get rid of __points - not hide it behind the wrapper func; it's same tomatoes with worse performance
b) there is no code duplication here, you are replacing one function call for another, one line of code for another
c) same reasons as a)

Reply:
a) Yes, would be nice, but in a separate PR that also addresses the whole tail of additional problems like resource duplication.
b) Having an existing function
```gdscript
func get_point_position(key: int) -> Vector2:
    return __points.get_point_position(key)
```
and then using `__points.get_point_position(key)` instead of `get_point_position(key)`, I would consider this code duplication, even though it's only 8 characters difference.
I understand the performance argument, but 0.1ms in some rare cases where you create a shape with over 1000 points, that's ridiculous and not worth the change. Especially since it's not the performance bottleneck of this addon.

c) Yes.

Reply:
During point drag, this function is called in a loop over hundreds of points every frame in several places. 

It's not only about performance, it's also about obscuring implementation details in a class that cares about implementation. Encapsulation already achieved in PointArray class. There's just literally no reason to do this, but 8 characters. And you think performance concerns are ridiculous comparing to this?

Reply:
I actually find it more obscure to use `__points.get_point_position()` instead of an already existing `get_point_position()` function because there are now two different access methods floating around and the next person trying to read or modify this code wonders why.
Also, the same principles of "don't repeat yourself" apply here as well, even if it's just 8 chars. Let's say you rename `get_point_position` in `SS2D_Point_Array`, you now have to rename it in various places than just in `shape_base::get_point_position`. Of course there is no particular reason to rename this function in this case, but it's just an example to make a point.

The bottleneck is not calling this function but `get_vertices()` being called multiple times per frame instead of using a cache.

I find it ridiculous to have a shape with more than 100 points and then complaining about 0.1ms performance loss. We're not even talking about interpolated points due to bezier curves, we're talking about over 100 manually placed points in one shape.

Let's say you have 1000 points and it will cost 0.1ms because auf 1000 "needless" function calls. Let's say we implemented a cache so that `get_vertices()` is only executed once. Then the total wasted time is 0.1ms. Who cares? You won't even notice that.

To retain a stable 60 FPS framerate in the editor while dragging a point, the script needs to run in less than 16ms. You could call `get_vertices()` about 160 times and still be fine. Obviously, a bit less in reality, because there's still other code doing stuff, but that's still a lot.

I'm happy to do performance optimizations, especially on this pugin, but not prematurely and not for rare cases where people decide 100+ points per shape is a sane decision, and only for 0.1ms improvement on that one shape.
If this would improve runtime performance every frame by 0.1ms, I wouldn't even argue. But since this only happens in the editor and is only relevant when having insane amounts of points, I prefer clean code.

Reply:
This last commit introduces nothing new, and obscures issues that my commits created. I made it a problem - I will solve it. If you will be willing to drop it - I will finalize the review. If not - it's your repo - you can do it yourself.

Reply:
> Now you are introducing the second way of doing it in just a few places.

I'm not introducing it, it's already there and can be reused. It's a helper function wrapping `__points.get_point_position()`. It serves as a public interface but also as a helper. You could also use `get_point_array().get_point_position()` from outside, but it's very tedious.
Not using that function is basically manually inlining that function, which makes no sense IMO. It's not that I replaced `__points.get_point_position` with `get_point_array().get_point_position`, which would really make no sense.

> This last commit introduces nothing new, and obscures issues that my commits created

It introduces nothing new that's right, but it uses existing helpers to access the underlying data. Especially shape_closed should use these functions as `__points` is not a direct member of it.
It streamlines the way these underlying properties are accessed to reduce confusion between `_points` and `__points` while reducing code duplication, even if it's technically just 8 chars. In reality it's not just 8 chars but a certain function call on a certain object.

`__points` and `_points` is not an issue you created, it's an issue that was already present. Without refactoring more of the code base to fix the complete tail of problems attached to it, that's the way to go.
Also, even if there would exist only one `_points`, I would still use those wrapper-helpers as far as possible, because that's what they're for.

> I'm sure I'm not the only one using it to model whole level terrain, you can easily go even over 1000 vertices

I highly doubt that it's a regular use case to have such large polygons. Especially since it makes it very hard to layout and iterate on your level design. At least in my experience.

Anyway, to come to a conclusion here. I'll change that one line in question to give you the 0.1ms per drag frame and 1ms on load.
I will not drop the whole last commit, but that one line.

Reply:
Done